# 마이크로태스크
프라미스 핸들러 .then/catch/finally는 항상 비동기적으로 실행된다.
프로미스가 즉이 이행되더라도 .then/catch/finally 아래에 있는 코드는 이 핸들러들이 실행되기전에 실행된다.
```java
let promise = Promise.resolve();
promise.then(() => alert('프라미스 성공!'));
alert('코드종료)' // 먼저 뜸
```

## 마이크로태스크 큐
비동기 작업을 처리하려면 적절한 관리가 필요하다. 이를 위해 ECMA에선 PromiseJobs라는 내부 큐를 명시한다. = 마이크로태스크 큐
* 먼저 들어온 작업을 먼저 실행한다(선입선출 FIFO)
* 실행할 것이 아무것도 남아있지 않을때만 마이크로태스크 큐에 있는 작업이 실행된다.
요약하자면 어떤 프라미스가 준비되었을때 이 프라미스의 .then/catch/finally 핸들러가 큐에 들어간다고 생각하면 된다.

### 처리되지 못한 거부 'unhandledrejection'
'처리되지 못한 거부'는 마이크로태스크 큐 끝에서 프라미스 에러가 처리되지 못할 때 발생한다.<br>
정상적인 경우라면 개발자는 에러가 생길 것을 대비하여 프로미스 체인에 .catch를 추가해 에러를 처리한다.
```java
let promise = Promise.reject(new Error('프라미스 실패 !'));
promise.catch(err => console.log('잡았다'))

window.addEventListener('unhandledrejection', event => console.log(event.name)) // catch를 선언 하였으므로 실행되지 않음.
```
'unhandledrejection'는 마이크로태스크 큐에 있는 작업 모두가 완료 되었을때 생성된다. 엔진은 프라미스들을 검사하고 이 중 하나라도
reject상태이면 unhandledrejection 핸들러를 트리거 한다.
```java
let promise = Promise.reject(new Error("프라미스 실패!"));
setTimeout(() => promise.catch(err => alert('잡았다!')), 1000);

// Error: 프라미스 실패!
window.addEventListener('unhandledrejection', event => alert(event.reason));
```
